# Data Science Unit 1 Sprint Challenge 4

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 (the class attribute) is the most interesting, as it separates the 307 approved cases from the 383 rejected cases. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [2]:
!curl -O https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data
!curl -O https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.names
!curl -O https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/credit.lisp
!curl -O https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/credit.names

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 32218  100 32218    0     0   170k      0 --:--:-- --:--:-- --:--:--  170k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1486  100  1486    0     0  12593      0 --:--:-- --:--:-- --:--:-- 12593
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 12314  100 12314    0     0  94000      0 --:--:-- --:--:-- --:--:-- 94000
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   522  100   522    0     0   4176      0 --:--:-- --:--:-- --:--:--  4176


In [3]:
!ls -l c*

-rw-r--r--  1 l  staff  12314 May 10 08:51 credit.lisp
-rw-r--r--  1 l  staff    522 May 10 08:51 credit.names
-rw-r--r--  1 l  staff  32218 May 10 08:51 crx.data
-rw-r--r--  1 l  staff   1486 May 10 08:51 crx.names


In [10]:
!cat credit.lisp

;; positive examples represent people that were granted credit
(def-pred credit_screening :type (:person) 
  :pos
  ((s1) (s2) (s4) (s5) (s6) (s7) (s8) (s9) (s14) (s15) (s17) (s18) (s19)
   (s21) (s22) (s24) (s28) (s29) (s31) (s32) (s35) (s38) (s40) (s41)
   (s42) (s43) (s45) (s46) (s47) (s49) (s50) (s51) (s53) (s54) (s55)
   (s56) (s57) (s59) (s61) (s62) (s63) (s64) (s65) (s66) (s69) (s70)
   (s71) (s72) (s73) (s74) (s75) (s76) (s77) (s78) (s79) (s80) (s81)
   (s83) (s84) (s85) (s86) (s87) (s89) (s90) (s91) (s92) (s93) (s94)
   (s96) (s97) (s98) (s100) (s103) (s104) (s106) (s108) (s110) (s116)
   (s117) (s118) (s119) (s121) (s122) (s123) (s124))
  :neg
  ((s3) (s10) (s11) (s12) (s13) (s16) (s20) (s23) (s25) (s26) (s27) 
   (s30) (s33) (s34) (s36) (s37) (s39) (s44) (s48) (s52) (s58) (s60)
   (s67) (s68) (s82) (s88) (s95) (s99) (s101) (s102) (s105) (s107)
   (s109) (s111) (s112) (s113) (s114) (s115) (s120) (s125)))

(def-pred jobless :type (:person) :pos
  ((s3) (s10) (s12) (s23) (s34) 

In [63]:
import pandas as pd

In [64]:
df = pd.read_csv('crx.data', header=None, na_values='?')

In [65]:
df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  202.0    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g   43.0  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  280.0  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  100.0    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  120.0    0  +

In [66]:
df[3].unique()

array(['u', 'y', nan, 'l'], dtype=object)

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? **Choose two features to run t-tests (again split by class label)** - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

**There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result** (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [67]:
df[15].unique()

array(['+', '-'], dtype=object)

In [68]:
approved = df.loc[df[15] == '+']
disappro = df.loc[df[15]=='-']

assert len(df)== len(approved) + len(disappro)

In [79]:
approved.isna().sum()

0     3
1     2
2     0
3     4
4     4
5     4
6     4
7     0
8     0
9     0
10    0
11    0
12    0
13    6
14    0
15    0
dtype: int64

In [81]:
disappro.isna().sum()

0      9
1     10
2      0
3      2
4      2
5      5
6      5
7      0
8      0
9      0
10     0
11     0
12     0
13     7
14     0
15     0
dtype: int64

In [69]:
approved.describe()

1           2           7           10          13  \
count  305.000000  307.000000  307.000000  307.000000  301.000000   
mean    33.720492    5.904951    3.427899    4.605863  164.421927   
std     12.809641    5.471485    4.120792    6.320242  161.770675   
min     13.750000    0.000000    0.000000    0.000000    0.000000   
25%     23.170000    1.500000    0.750000    0.000000    0.000000   
50%     30.500000    4.460000    2.000000    3.000000  120.000000   
75%     41.330000    9.520000    5.000000    7.000000  280.000000   
max     76.750000   28.000000   28.500000   67.000000  840.000000   

                  14  
count     307.000000  
mean     2038.859935  
std      7659.763941  
min         0.000000  
25%         0.000000  
50%       221.000000  
75%      1209.000000  
max    100000.000000

In [70]:
disappro.describe()

1           2           7           10           13  \
count  373.000000  383.000000  383.000000  383.000000   376.000000   
mean    29.808231    3.839948    1.257924    0.631854   199.699468   
std     10.919291    4.337662    2.120481    1.900049   181.564835   
min     15.170000    0.000000    0.000000    0.000000     0.000000   
25%     22.000000    0.835000    0.125000    0.000000   100.000000   
50%     27.330000    2.210000    0.415000    0.000000   167.500000   
75%     34.830000    5.000000    1.500000    0.000000   272.000000   
max     80.250000   26.335000   13.875000   20.000000  2000.000000   

                14  
count   383.000000  
mean    198.605744  
std     671.608839  
min       0.000000  
25%       0.000000  
50%       1.000000  
75%      67.000000  
max    5552.000000

In [93]:
import scipy.stats

In [105]:
columns = [1, 2, 7, 10, 13, 14]
for elem in columns:
    print(elem, ':', ttest_ind(approved[elem], disappro[elem]))

1 : Ttest_indResult(statistic=nan, pvalue=nan)
2 : Ttest_indResult(statistic=5.52998337614816, pvalue=4.551680702308068e-08)
7 : Ttest_indResult(statistic=8.935819983773698, pvalue=3.6710537401601785e-18)
10 : Ttest_indResult(statistic=11.667004222431277, pvalue=7.957718568079967e-29)
13 : Ttest_indResult(statistic=nan, pvalue=nan)
14 : Ttest_indResult(statistic=4.680216020964486, pvalue=3.4520256956287944e-06)


In [106]:
columns = [1, 2, 7, 10, 13, 14]
for elem in columns:
    print(elem, ':', ttest_ind(approved.dropna()[elem], disappro.dropna()[elem]))

1 : Ttest_indResult(statistic=4.675662433767456, pvalue=3.5636710933835225e-06)
2 : Ttest_indResult(statistic=5.400813416576192, pvalue=9.310154396147606e-08)
7 : Ttest_indResult(statistic=9.002392498622463, pvalue=2.4079238505859142e-18)
10 : Ttest_indResult(statistic=11.336964562239809, pvalue=2.5864649620278843e-27)
13 : Ttest_indResult(statistic=-2.18221936438708, pvalue=0.029450100409286473)
14 : Ttest_indResult(statistic=4.475369764700449, pvalue=9.003915641872878e-06)


In [74]:
import numpy as np

In [75]:
approved.describe(exclude=np.number)

0    3    4    5    6    8    9    11   12   15
count   304  303  303  303  303  307  307  307  307  307
unique    2    3    3   14    9    2    2    2    3    1
top       b    u    g    c    v    t    t    f    g    +
freq    206  256  256   62  169  284  209  161  287  307

In [76]:
disappro.describe(exclude=np.number)

0    3    4    5    6    8    9    11   12   15
count   374  381  381  378  378  383  383  383  383  383
unique    2    2    2   14    9    2    2    2    3    1
top       b    u    g    c    v    f    f    f    g    -
freq    262  263  263   75  230  306  297  213  338  383

In [110]:
columns = [0, 3, 4, 5, 6, 8, 9, 11, 12]
for subject in columns:
    ct = pd.crosstab(df[15], df[subject])
    chi_sqr, p_value, dof, expected = scipy.stats.chi2_contingency(ct)
    print(subject, ':  chisqr =', chi_sqr,'  p_val =', p_value)

0 :  chisqr = 0.31128326491619945   p_val = 0.5768937883001117
3 :  chisqr = 26.234074966202144   p_val = 2.010680204180363e-06
4 :  chisqr = 26.234074966202144   p_val = 2.010680204180363e-06
5 :  chisqr = 98.32520342679135   p_val = 3.4999300402715717e-15
6 :  chisqr = 45.034207140240554   p_val = 3.625452872372267e-07
8 :  chisqr = 355.2038167412799   p_val = 3.1185900878457007e-79
9 :  chisqr = 143.06956205083145   p_val = 5.675727374527571e-33
11 :  chisqr = 0.568273300792113   p_val = 0.45094587758631943
12 :  chisqr = 9.191570451545383   p_val = 0.010094291370456362


## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

### T-Tests:

In [103]:
ttest_ind(approved[10], disappro[10])

Ttest_indResult(statistic=11.667004222431277, pvalue=7.957718568079967e-29)

Column 10 has a t-stat of 11.667, or a .0000000000000000000000000001% chance of approval being independent of column 10. This can also be read as a 99.99999999999999999999999999% confidence level. I'm confident in my assertion that column 10 does strongly affect the outcome of approval.

In [107]:
ttest_ind(approved.dropna()[13], disappro.dropna()[13])

Ttest_indResult(statistic=-2.18221936438708, pvalue=0.029450100409286473)

Column 13 has a t-stat of -2.182293. This disproves the Null hypothesis that column 13 does not effect approval. This is however only at a 97.05498995907136% Confidence level. This is still above the traditionally used standard of 95%. 

## Chi_squared tests:

In [111]:
subject = 8
chi_sqr, p_value, dof, expected = scipy.stats.chi2_contingency(pd.crosstab(df[15], df[subject]))
chi_sqr, p_value

(355.2038167412799, 3.1185900878457007e-79)

Column 13 has a chi_sqr value of 355. This shows near certainty that Approval is dependent on this value. The probability that this is from chance is at 3.1185900878457007e-79. 

In [112]:
subject = 0
chi_sqr, p_value, dof, expected = scipy.stats.chi2_contingency(pd.crosstab(df[15], df[subject]))
chi_sqr, p_value

(0.31128326491619945, 0.5768937883001117)

Column 0 has a chi-sqr value of .311. This is not enough to disprove the Null hypothesis, it has a 60% probability of the results being simply due to chance, and not the column 0 stat.

## Final question:

The hardest part of this sprint challenge was describing what our results meant. It was challenging because you had to beable to not only understand the results, but be able to describe them in a simple way.